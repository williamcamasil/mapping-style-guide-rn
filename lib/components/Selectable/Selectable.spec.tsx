import React from 'react';

import { render, fireEvent } from '@testing-library/react-native';

import { SelectableItemType } from './index';

import { Selectable } from '..';

type Item = {
  id: number;
  name: string;
};

const items: SelectableItemType<Item>[] = [
  {
    key: 1,
    label: 'Conta corrente / Movimentação',
    item: { id: 1, name: 'Conta corrente / Movimentação' },
  },
  {
    key: 2,
    label: 'Investimento',
    item: { id: 2, name: 'Investimento' },
  },
  {
    key: 3,
    label: 'Operação de Crédito',
    item: { id: 3, name: 'Operação de Crédito' },
  },
  {
    key: 4,
    label: 'Produtos / Serviços',
    item: { id: 4, name: 'Produtos / Serviços' },
  },
  {
    key: 5,
    label: 'Outros',
    item: { id: 5, name: 'Outros' },
  },
];

describe('Selectable', () => {
  it('Should render all items, with none selected', () => {
    const jsonScreen = render(<Selectable options={items} onChange={jest.fn()} value={[]} />).toJSON();
    expect(jsonScreen).toMatchSnapshot();
  });

  it('Should render all items, with the first two selected', () => {
    const jsonScreen = render(<Selectable
      options={items}
      onChange={jest.fn()}
      value={[
        {
          key: 1,
          label: 'Conta corrente / Movimentação',
          item: { id: 1, name: 'Conta corrente / Movimentação' },
        },
        {
          key: 2,
          label: 'Investimento',
          item: { id: 2, name: 'Investimento' },
        },
      ]}
    />).toJSON();
    expect(jsonScreen).toMatchSnapshot();
  });

  it('Should call onChange when one of the items is selected', async () => {
    const onChange = jest.fn();
    const screen = render(<Selectable options={items} onChange={onChange} value={[]} />);
    onChange.mockClear();
    const button = await screen.findByTestId('selectable-button-item-3');
    fireEvent.press(button);

    expect(onChange).toHaveBeenCalledWith([
      {
        key: 4,
        label: 'Produtos / Serviços',
        item: { id: 4, name: 'Produtos / Serviços' },
      },
    ]);
  });

  it('Should deselect the item when clicked after it is already selected', async () => {
    const onChange = jest.fn();
    const screen = render(
      <Selectable
        options={items}
        onChange={onChange}
        value={[
          {
            key: 1,
            label: 'Conta corrente / Movimentação',
            item: { id: 1, name: 'Conta corrente / Movimentação' },
          },
          {
            key: 2,
            label: 'Investimento',
            item: { id: 2, name: 'Investimento' },
          },
        ]}
      />,
    );
    const button = await screen.findByTestId('selectable-button-item-0');
    fireEvent.press(button);

    expect(onChange).toHaveBeenLastCalledWith([
      {
        key: 2,
        label: 'Investimento',
        item: { id: 2, name: 'Investimento' },
      },
    ]);
  });
});
