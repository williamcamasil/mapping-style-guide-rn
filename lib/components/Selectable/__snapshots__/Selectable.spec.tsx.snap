// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectable Should render all items, with none selected 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
  testID="selectable-button-group"
>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-0"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Conta corrente / Movimentação
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-1"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Investimento
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-2"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Operação de Crédito
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-3"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Produtos / Serviços
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-4"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Outros
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Selectable Should render all items, with the first two selected 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
  testID="selectable-button-group"
>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#C6D3EA",
            "borderRadius": 12,
            "borderWidth": 0,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-0"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#013188",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#013188",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Conta corrente / Movimentação
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#C6D3EA",
            "borderRadius": 12,
            "borderWidth": 0,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-1"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#013188",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#013188",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Investimento
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-2"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Operação de Crédito
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-3"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Produtos / Serviços
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 12,
        "marginRight": 12,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
      testID="container-button"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#E1E2E4",
            "borderRadius": 12,
            "borderWidth": 1,
            "display": "flex",
            "height": 48,
            "justifyContent": "center",
            "minWidth": 48,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="selectable-button-item-4"
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#676D79",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 19,
            }
          }
          testID="text-with-icon"
        >
          <Text
            allowFontScaling={false}
            style={
              {
                "color": "#676D79",
                "flexShrink": 1,
                "fontFamily": "Averta CY",
                "fontSize": 13,
                "fontWeight": "bold",
                "letterSpacing": -0.02,
                "lineHeight": 17,
              }
            }
          >
            Outros
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
